apps/prairielearn/src/middlewares/authzWorkspace.ts

```diff
@@ -60,6 +60,12 @@ export default asyncHandler(async (req, res, next) => {
       await authzHasCoursePreviewOrInstanceView(req, res);
       await selectAndAuthzInstructorQuestion(req, res);
     }
+  } else if (res.locals.share_publicly || res.locals.share_source_publicly) {
+    // This workspace is not part of a course instance and the question is publicly accessible
+    // Therefore, just make sure its being accessed by the same person who created the variant
+    if (res.locals.authn_user.user_id !== res.locals.variant_user_id) {
+      throw new HttpStatusError(403, 'Access denied');
+    }
   } else if (res.locals.course_id) {
```

Comment:
I just realized another problem here, which is that if a question is publicy shared, but is being looked at by an instructor on a preview within their course, we actually don't want to use the new auth method, we want it to still use the auth method under `else if (res.locals.course_id)`

So we need a better way to tell which URL a workspace questions is getting access through, rather than just checking for the existence of certain things on res.locals.

Reply:
For some other endpoints, like client files, we have a public version of the endpoint, then we can pass in a 'public' parameter to the handler. 

Would that be appropriate to do here as well?

Reply:
Does it actually make a difference? If a user is accessing a workspace associated to a variant created by them, would there be a case where they would no longer get access to that variant? The only situation I can think of that would cause this to happen is a user that used to be staff in a course and no longer is staff, but considering this is a publicly shared question, the user could presumably access the same variant by crafting a well-defined public URL. In fact, I just tested this:

* In a publicly accessible question, I opened the preview (not public) of the question in a course URL. This created a variant.
* I submitted an answer to the variant. This redirected to the preview page with the variant_id query parameter in the URL.
* I opened the public version of the same question, and appended the same variant_id query parameter to the URL. This resulted in the previous variant being loaded without problems, even though it was created outside of the public context.

What I'm trying to say is: we currently don't make a distinction between variants created in a course route in preview and variants created in public preview, so we should make this distinction in how the authn for this is handled, as long as there is no course instance ID in the variant.

With that being said: I believe the comparison should be done with `v.authn_user_id`, not `v.user_id`, but I don't have a strong opinion in that matter.

Reply:
Also: probably a good idea to review what authzCourseOrInstance is updating, to make sure any locals updated there and needed in the workspace context (or affecting it's functionality/rendering) are properly updated as needed in this context as well.

Reply:
> review what authzCourseOrInstance is updating, to make sure any locals updated there and needed in the workspace context

Yes, this is the problem I was referring to (sorry I wasn't more specific). 

If the call to `authzCourseOrInstance` is not made, then ` res.locals.authz_data` doesn't get set, and then in the call to `Workspace` (in apps/prairielearn/src/pages/workspace/workspace.ts), then in the case where an instructor is viewing a workspace inside their course, when the question happens to also be publicly shared, then the `showLogs` parameter will evaluate to false when it should actually be true